tools/migrate_kv.ts

```diff
@@ -1,14 +1,140 @@
 // Copyright 2023 the Deno authors. All rights reserved. MIT license.
-import { kv, type Notification } from "@/utils/db.ts";
+import {
+  assertIsEntry,
+  Item,
+  kv,
+  newVoteProps,
+  type Notification,
+  User,
+  Vote,
+} from "@/utils/db.ts";
+
+type OldItem = Omit<Item, "userLogin"> & { userId: string };
+type OldUser = User & { id: string };
+
+async function migrateVote(voterUserId: string, oldItem: OldItem) {
+  const [creatorUserRes, voterUserRes] = await kv.getMany<OldUser[]>([
+    ["users", oldItem.userId],
+    ["users", voterUserId],
+  ]);
+  assertIsEntry(creatorUserRes);
+  assertIsEntry(voterUserRes);
+
+  const item: Item = {
+    userLogin: creatorUserRes.value.login,
+    id: oldItem.id,
+    createdAt: oldItem.createdAt,
+    title: oldItem.title,
+    url: oldItem.url,
+    score: oldItem.score,
+  };
+  const vote: Vote = {
+    ...newVoteProps(),
+    item,
+    userLogin: voterUserRes.value.login,
+  };
+
+  const itemKey = ["items", vote.item.id];
+  const itemsByTimeKey = [
+    "items_by_time",
+    vote.item.createdAt.getTime(),
+    vote.item.id,
+  ];
+  const itemsByUserKey = ["items_by_user", vote.item.userLogin, vote.item.id];
+  const [itemRes, itemsByTimeRes, itemsByUserRes] = await kv.getMany([
+    itemKey,
+    itemsByTimeKey,
+    itemsByUserKey,
+  ]);
+  assertIsEntry(itemRes);
+  assertIsEntry(itemsByTimeRes);
+  assertIsEntry(itemsByUserRes);
+
+  const votesKey = ["votes", vote.id];
+  const votesByItemKey = ["votes_by_item", vote.item.id, vote.id];
+  const votesByUserKey = ["votes_by_user", vote.userLogin, vote.id];
+
+  const res = await kv.atomic()
+    .check(itemRes)
+    .check(itemsByTimeRes)
+    .check(itemsByUserRes)
+    .set(itemKey, vote.item)
+    .set(itemsByTimeKey, vote.item)
+    .set(itemsByUserKey, vote.item)
+    .set(votesKey, vote)
+    .set(votesByItemKey, vote)
+    .set(votesByUserKey, vote)
+    .commit();
+
+  if (!res.ok) throw new Error(`Failed to set vote: ${vote}`);
+}
+
+async function migrateItem(oldItem: OldItem) {
```

Comment:
This is run after `migrateVote()` and only takes care of those items with no votes.